web3.js/src/transaction.js

```diff
@@ -210,54 +210,79 @@ export class Transaction {
     let numReadonlySignedAccounts = 0;
     let numReadonlyUnsignedAccounts = 0;
 
-    const accountKeys = this.signatures.map(({publicKey}) =>
-      publicKey.toString(),
-    );
-    const programIds: string[] = [];
+    const programIds: PublicKey[] = [];
     const accountMetas: AccountMeta[] = [];
     this.instructions.forEach(instruction => {
       instruction.keys.forEach(accountMeta => {
         accountMetas.push(accountMeta);
       });
 
-      const programId = instruction.programId.toString();
+      const programId = instruction.programId;
       if (!programIds.includes(programId)) {
         programIds.push(programId);
       }
     });
 
+    // Append programID account metas
+    programIds.forEach(programId => {
+      accountMetas.push({
+        pubkey: programId,
+        isSigner: false,
+        isWritable: false,
+      });
+    });
+
+    // Prefix accountMetas with feePayer here whenever that gets implemented
```

Comment:
The lack of this is your issue.

Need something like this, I imagine:
```
this.signatures.forEach(signature => {
      if (!accountMetas.find(meta => meta.pubkey.toString() === signature.publicKey.toString())) {
        accountMetas.unshift({
          pubkey: signature.publicKey,
          isSigner: true,
          isWritable: true,
        });
      }
    });
```

Reply:
Nice catch! :pray:

Reply:
I went with https://github.com/solana-labs/solana/pull/10950/commits/1e11573802be58036527c29bf40adf79221d2b15#diff-86eea70a26ecf01c3229f804beb21ac4R260-R272